Added :meth:`.set_opacity_by_tex` method for setting the opacity of parts of Tex mobjects

MatthewLeeCode opened the pull request:
<!-- Thank you for contributing to Manim! Learn more about the process in our contributing guidelines: https://docs.manim.community/en/latest/contributing.html -->

## Overview: What does this pull request change?
<!-- If there is more information than the PR title that should be added to our release changelog, add it in the following changelog section. This is optional, but recommended for larger pull requests. -->

Add two functions:
`MathTex.set_opacity_by_tex`: Matches naming style of `MathTex.set_color_by_tex`
`MathTex.fade_all_but_tex`: Matches naming style of `BulletedList.fade_all_but`

<!--changelog-start-->

<!--changelog-end-->

## Motivation and Explanation: Why and how do your changes improve the library?
<!-- Optional for bugfixes, small enhancements, and documentation-related PRs. Otherwise, please give a short reasoning for your changes. -->

Ability to set the opacity of specific parts of the latex string. This can be used to add emphasis to sections as discussing them. This is currently possible with the following code:
```Python
tex = MathTex("f(x) = y^2 + x", substrings_to_isolate=["f(x)"])

tex_part = tex.get_part_by_tex("f(x)")
tex_part.set_opacity(0.2)
```

With change:
```Python
tex = MathTex("f(x) = y^2 + x", substrings_to_isolate=["f(x)"])

tex.set_opacity_by_tex("f(x)", 0.2)
```
![set_opacity_by_tecx](https://user-images.githubusercontent.com/12680255/219247667-7c818dda-ae1f-40a9-b756-de15a664703c.png)

This matches how we use `MathTex.set_color_by_tex`. 

Additionally, the function 'MathTex.fade_all_but_tex` is added to fade all parts except the tex specified.
```Python
tex = MathTex("f(x) = y^2 + x", substrings_to_isolate=["f(x)"])

tex.fade_all_but_tex("f(x)", 0.2)
self.add(tex)
```
![fade_all_but_tex](https://user-images.githubusercontent.com/12680255/219247627-c1d1893e-1064-456a-8c49-5c8d4dc5c415.png)

This all works with 'animate' as well to fade in and out

<!-- Thank you again for contributing! Do not modify the lines below, they are for reviewers. -->
## Reviewer Checklist
- [ ] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

MatthewLeeCode commented:
Hey there. It appears I have messed something up with the checks. I thought I followed everything. Very sorry. I'll see if I can resolve this

behackl reviewed (changes requested):
Thanks for your contribution!

I agree that this addition makes sense. I do wonder though, whether it really has to be two separate methods though, or whether `set_opacity_by_tex` could just also get a second opacity argument for the non-matched values; the implementation could then simply be: set the entire string to the opacity for the non-matched values (or leave it as is if `None`) is passed, then set the opacity of the mached parts.

If I am not mistaken, the `fade_all_but_text` implementation has the problem that the match will only be found (and not faded) a single time, regardless of how often it occurs in the string.

Therefore, my proposal would be: incoprorate the functionality of `fade_all_but_tex` in `set_opacity_by_tex`, and add a docstring to `set_opacity_by_tex` to explain the keyword arguments.

Let me know what you think!

MatthewLeeCode commented:
Thanks @behackl ! I completely agree. I have added the functionality to `set_opacity_by_tex` and removed the function `fade_all_but_tex`. The second parameter is referred to as 'remaining_opacity' which may or may not be a good name. I'm happy to change it if you'd prefer a different name for it. Of course the hardest part of any programming implementation is naming things...

Let me know your thoughts!

behackl commented on manim/mobject/text/tex_mobject.py:
```diff
@@ -381,6 +381,29 @@ def set_color_by_tex(self, tex, color, **kwargs):
+            Default None. The opacity to set the remaining tex to.
+            If None, then the remaining tex will not be changed
+        """
+        part_index = self.index_of_part_by_tex(tex)
```
With this approach, you will only be able to select the specified tex (sub)string a single time, further occurrences are not considered (as `index_of_part_by_tex` only returns the index of the first match).

You could change this to the previous implementation by doing something like this:
```
if remaining_opacity is not None:
    self.set_opacity(opacity=remaining_opacity)
for part in self.get_parts_by_tex(...):
    part.set_opacity(opacity)
```

What do you think?

behackl commented on manim/mobject/text/tex_mobject.py:
```diff
@@ -381,6 +381,29 @@ def set_color_by_tex(self, tex, color, **kwargs):
+
+        Parameters
+        ----------
+        tex : str
+            The tex to set the opacity of.
+        opacity : float, optional
+            Default 0.5. The opacity to set the tex to
+        remaining_opacity : float, optional
```
The type hints should be moved from the docstring to the method head: `tex` -> `tex: str`, `opacity` -> `opacity: float = 0.5`, etc.


```suggestion
        tex
            The tex to set the opacity of.
        opacity
            Default 0.5. The opacity to set the tex to
        remaining_opacity
```

MatthewLeeCode replied:
Sorry it took me a while to get to this. I agree with both changes mentioned. Unfortunately, the auto-formatter that runs during commit moved the parameters onto their own line. Not a fan of this personally but it is required to pass the checks. Let me know what you think of the changes. Thanks!

behackl reviewed (changes requested):
Sorry for the late reply! Two further comments, and one suggestion to improve the implementation.

behackl commented on manim/mobject/text/tex_mobject.py:
```diff
@@ -381,6 +381,29 @@ def set_color_by_tex(self, tex, color, **kwargs):
         return self
 
+    def set_opacity_by_tex(
+        self, tex: str, opacity: float = 0.5, remaining_opacity: float = None, **kwargs
```
Minor type hint mismatch

```suggestion
        self, tex: str, opacity: float = 0.5, remaining_opacity: float | None = None, **kwargs
```

behackl commented on manim/mobject/text/tex_mobject.py:
```diff
@@ -381,6 +381,29 @@ def set_color_by_tex(self, tex, color, **kwargs):
+    def set_opacity_by_tex(
+        self, tex: str, opacity: float = 0.5, remaining_opacity: float = None, **kwargs
+    ):
+        """
+        Sets the opacity of the tex specified. If 'remaining_opacity' is specified,
```
Just for consistency, and formatting

```suggestion
        """Sets the opacity of the tex specified. If ``remaining_opacity`` is specified,
```

behackl reviewed (approved):
Thanks @MatthewLeeCode, and sorry for the slow reply (again). I'm fine with the proposed changes as-is, I've just left a few suggestions regarding wording/formatting of the docstring.

Let me know if you find a moment to look at the suggestions and either agree with them, or whether you'd like to have something changed -- otherwise I'll merge this in a day or two so that this is included in the next release.

Thanks again for your contribution!

MatthewLeeCode commented:
Thanks @behackl appreciate your time. Agree with the suggestions and looking forward to seeing this in the next release :)
